sql/sql_select.cc

```diff
@@ -8050,8 +8050,13 @@ best_access_path(JOIN      *join,
             }
 
             /* Limit the number of matched rows */
-            tmp= cost_for_index_read(thd, table, key, (ha_rows) records,
-                                     (ha_rows) s->worst_seeks);
+            ha_rows worst_seeks; 
+            if (isnan(s->worst_seeks))
+              worst_seeks = (ha_rows) UINT_MAX;
+            else
+              worst_seeks = (ha_rows) s->worst_seeks;
+            tmp= cost_for_index_read(thd, table, key,
+                                     (ha_rows) records, worst_seeks);
```

Comment:
It must be oneliner like before. `cost_for_index_read()` always accepts double as `worst_seeks`. It's signature should be changed to `double` and NaN handling should be done inside `cost_for_index_read()`. 

I'd suggest to add something like:

```c++
inline ha_rows
double2ha_rows(double n)
{
  return isnan(n) ? (ha_rows) HA_ROWS_MAX : (ha_rows) n;
}
```
(using `UINT_MAX` is wrong as `HA_ROWS` can be `ulonglong`).